### PR TITLE
net/local: add AF_[UNIX|LOCAL] socketpair support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/local: add AF_[UNIX|LOCAL] socketpair support
net/socketpair: move socketpair implement into socket internal 
net/local: extern local_generate_instance_id() interface 

## Impact

socketpair wirth AF_UNIX  /  AF_LOCAL  / AF_INET

## Testing

socketpair wirth AF_UNIX and AF_LOCAL

### CODE REVIEW

File: net/socket/socketpair.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+#include <nuttx/kmalloc.h>
+#include <nuttx/net/net.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: socketpair
```
Comment (xiaoxiang781216) [inline net/socket/socketpair.c:42]: psock_socketpair

File: net/inet/inet_sockif.c [hunk 2]
```diff
@@ -1328,6 +1342,147 @@ static int inet_ioctl(FAR struct socket *psock, int cmd,
...
+      if (ret < 0)
+        {
+          return ret;
+        }
+
+      pserver = &server;
+    }
+#endif /* CONFIG_NET_TCP */
+
+#if defined(CONFIG_NET_UDP)
```
Comment (xiaoxiang781216) [inline net/inet/inet_sockif.c:1415]: else
#endif
  {

File: net/inet/inet_sockif.c [hunk 3]
```diff
@@ -1328,6 +1342,147 @@ static int inet_ioctl(FAR struct socket *psock, int cmd,
...
+
+      ret = psock_connect(pserver, &addr[0].addr, len);
+      if (ret < 0)
+        {
+          goto errout;
+        }
+    }
+#endif /* CONFIG_NET_UDP */
+
+#if defined(CONFIG_NET_TCP)
```
Comment (xiaoxiang781216) [inline net/inet/inet_sockif.c:1447]: else
  {
#endif

File: net/inet/inet_sockif.c [hunk 4]
```diff
@@ -1328,6 +1342,147 @@ static int inet_ioctl(FAR struct socket *psock, int cmd,
...
+    }
+
+  psock_getsockname(psocks[0], &addr[0].addr, &len);
+
+  /* For SOCK_STREAM, Use proxy service handle to make temporary
+   * pserver process, psocks[1] will be replaced with a new accept handle
+   */
+
+#if defined(CONFIG_NET_TCP)
+  FAR struct socket server;
```
Comment (xiaoxiang781216) [inline net/inet/inet_sockif.c:1400]: move to the begin of function

File: net/local/local_sockif.c [hunk 5]
```diff
@@ -728,6 +731,106 @@ static int local_ioctl(FAR struct socket *psock, int cmd,
...
+  bool nonblock;
+  int ret;
+#endif /* CONFIG_NET_LOCAL_STREAM */
+  int i;
+
+  for (i = 0; i < 2; i++)
+    {
+      conns[i] = psocks[i]->s_conn;
+      snprintf(conns[i]->lc_path,
+               sizeof(conns[i]->lc_path), "%p", psocks[0]);
```
Comment (xiaoxiang781216) [inline net/local/local_sockif.c:759]: socketpair%p

File: net/local/local_sockif.c [hunk 6]
```diff
@@ -728,6 +731,106 @@ static int local_ioctl(FAR struct socket *psock, int cmd,
...
+
+      conns[i]->lc_proto = psocks[i]->s_type;
+      conns[i]->lc_type  = LOCAL_TYPE_PATHNAME;
+      conns[i]->lc_state = LOCAL_STATE_BOUND;
+    }
+
+#ifdef CONFIG_NET_LOCAL_DGRAM
+#ifdef CONFIG_NET_LOCAL_STREAM
```
Comment (xiaoxiang781216) [inline net/local/local_sockif.c:767]: remove?

File: net/local/local_sockif.c [hunk 7]
```diff
@@ -728,6 +731,106 @@ static int local_ioctl(FAR struct socket *psock, int cmd,
...
+  if (psocks[0]->s_type == SOCK_DGRAM)
+#endif /* CONFIG_NET_LOCAL_STREAM */
+    {
+      return OK;
+    }
+#endif /* CONFIG_NET_LOCAL_DGRAM */
+
+#ifdef CONFIG_NET_LOCAL_STREAM
```
Comment (xiaoxiang781216) [inline net/local/local_sockif.c:775]: #else

File: net/inet/inet_sockif.c (hunk 2 above)
Comment (anchao) [reply net/inet/inet_sockif.c:1415]: TCP / UDP is a completely independent scope. which can be enabled separately, so it is impossible to add else to above scope

File: net/local/local_sockif.c (hunk 5 above)
Comment (anchao) [reply net/local/local_sockif.c]: Done

File: net/local/local_sockif.c (hunk 6 above)
Comment (anchao) [reply net/local/local_sockif.c:767]: if define the UDP without TCP, remove this ifdef will caused compile warning "no return value"

**Review by xiaoxiang781216 — APPROVED** [review]

File: net/socket/socketpair.c [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+  return OK;
+
+errout_with_psock:
+  while (i-- > 0)
+    {
+      nx_close(sv[i]);
+    }
+
+  i = 2;
```
Comment (yamt) [inline net/socket/socketpair.c:171]: doesn't this cause double-free of psock for which you did nx_close above?

**Review by yamt — CHANGES_REQUESTED** [review]

File: net/socket/socketpair.c (hunk 8 above)
Comment (anchao) [reply net/socket/socketpair.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM, @yamt could you take a look?

**Pull request merged**